README.md
# ![Magickly - image manipulation as a (plugin-able) service](http://magickly.jux.com/images/logo.jpg)

Built as a practical wrapper of Imagemagick which handles caching, c/o the [Dragonfly](http://markevans.github.com/dragonfly/) gem.

Say the base URL is the hosted version of this app, [magickly.jux.com](http://magickly.jux.com).  The image URL is appended to the query string as a `src=`, followed by any of the supported operations below.  Multiple operations can be combined, and will be applied in order.

If no query params are provided, a simple sandbox page is displayed.  Try it here:

[magickly.jux.com](http://magickly.jux.com)

Blog post about how it's used at Jux:

[aidan.jux.com/nerdery/310516](https://aidan.jux.com/nerdery/310516)

## Installation

[Compatible](http://travis-ci.org/#!/afeld/magickly) with Ruby 1.8.7, 1.9.2 and 1.9.3.  Requires Imagemagick >= v6.2.4.

    $ gem install magickly

## Running the App

A few options:

### A. Run the app directly

    # in the app directory:
    $ rackup

The app can be accessed at [http://localhost:9292](http://localhost:9292).  To deploy to Heroku's Cedar stack (or another server using Foreman), see the [cedar](https://github.com/afeld/magickly/tree/cedar) branch.

### B. Use as an endpoint in another Rack app

As an example, to have magickly accessible at `/magickly` in a Rails app:

    # Gemfile
    gem 'magickly', '~> 1.1'
    
    # config/routes.rb
    match '/magickly', :to => Magickly::App, :anchor => false

For more info, see [Rails Routing from the Outside In](http://guides.rubyonrails.org/routing.html#routing-to-rack-applications) or Michael Raidel's [Mount Rails apps in Rails 3](http://inductor.induktiv.at/blog/2010/05/23/mount-rack-apps-in-rails-3/).

## Processing Parameters

*See the [Dragonfly documentation](http://markevans.github.com/dragonfly/file.ImageMagick.html) for more details about the permitted* `geometry` *values.*

### src=*url* (required)

The URL of the original image.

### brightness_contrast=*br.* x *con.*

*brightness* and *contrast* are percentage change, between -100 and 100.  For example, to increase contrast by 20% but leave brightness unchanged, use `brightness_contrast=0x20`.

![tanned imagemagick logo](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&brightness_contrast=-10x50)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&brightness_contrast=-10x50](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&brightness_contrast=-10x50)

### flip=true

![flipped imagemagick logo](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&flip=true)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&flip=true](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&flip=true)

### flop=true

![flopped imagemagick logo](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&flop=true)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&flop=true](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&flop=true)

### glow=*amount*,*softness*

where `amount` is a float >= 1.0, and `softness` is an int >= 0.

![glowing imagemagick logo](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&glow=1.2,20)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&glow=1.2,20](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&glow=1.2,20)

### greyscale=true

![flopped imagemagick logo](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&greyscale=true)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&greyscale=true](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&greyscale=true)

### halftone=*threshold*

where *threshold* is a value between 0 and 100.

![halftone imagemagick logo](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&halftone=60)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&halftone=60](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&halftone=60)

### jcn=true

![JCN imagemagick logo](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&jcn=true)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&jcn=true](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&jcn=true)

### resize=*geometry*

![resized imagemagick logo](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&resize=100x100)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&resize=100x100](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&resize=100x100)

### rotate=*degrees*

![rotated imagemagick logo](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&rotate=45)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&rotate=45](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&rotate=45)

### saturation=*percentage*

*percentage* is the percentage of variation: a positive integer.  100 means no change.  For example, to increase saturation by 50%, use `saturation=150`.

![saturated imagemagick logo](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&saturation=150)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&saturation=150](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&saturation=150)

### tilt_shift=true

![tilt-shifted imagemagick logo](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&tilt_shift=true)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&tilt_shift=true](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&tilt_shift=true)

### thumb=*geometry*

![thumbnail of imagemagick logo](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&thumb=200x100%23)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&thumb=200x100%23](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&thumb=200x100%23)

(note: the `%23` in the geometry string above is an encoded '`#`', which tells Dragonfly to fill the dimensions and crop).  There is also an added geometry for doing cropping with a center: `WIDTHxHEIGHT#FOCUSX,FOCUSY`, where `FOCUSX` and `FOCUSY` are a percentage from the left and top respectively, between 0 and 100.  The photo will be resized and then cropped, trying to get that focus point as close to center as possible without leaving a gap.  For example if you wanted to crop around the wand, you would provide a focus point of `30,20`:

![thumbnail with focus of imagemagick logo](/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&thumb=300x100%2330,20)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&thumb=300x100%2330,20](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&thumb=300x100%2330,20)

### two_color=true

![two color imagemagick logo](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&two_color=true)

[http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&two_color=true](http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&two_color=true)

## Alternate Syntax

Some CDNs are jerks and don't respect query params on resources (_ahem_ CLOUDFRONT _ahem_) when caching.  To use this syntax:

* replace the question mark that starts the query string (`?`) with `q/`
* replace the ampersands (`&`) and equals signs (`=`) with forward slashes (`/`)
* make sure the `src` is encoded - this can be done in Javascript with `encodeURIComponent()`

Therefore, instead of

	http://magickly.jux.com/?src=http://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Imagemagick-logo.png/200px-Imagemagick-logo.png&thumb=200x100
	
the new URL would be

	http://magickly.jux.com/q/src/http%3A%2F%2Fupload.wikimedia.org%2Fwikipedia%2Fcommons%2Fthumb%2F0%2F0d%2FImagemagick-logo.png%2F200px-Imagemagick-logo.png/thumb/200x100

## Analyzers

Magickly v1.2.0 introduces the ability to retrieve image properties via a REST API.  For example, to retrieve the number of colors in the photo, visit:

[magickly.jux.com/analyze/number_of_colors?src=...](http://magickly.jux.com/analyze/number_of_colors?src=http://upload.wikimedia.org/wikipedia/commons/0/0d/Imagemagick-logo.png)

To get the list of available analyzers, visit [magickly.jux.com/analyze](http://magickly.jux.com/analyze)

## Customization

In addition to the available parameters listed above, custom "shortcuts" can be created to perform arbitrary imagemagick operations.  For example, to create a shortcut called `resize_with_blur`:

    # somewhere in your app configuration, i.e. config/initializers/magickly.rb for a Rails 3 app
    Magickly.dragonfly.configure do |c|
      c.job :resize_with_blur do |size|
        process :convert, "-filter Gaussian -resize #{size}"
      end
    end

which can then be used with the query string `?src=...&resize_with_blur=200x`.  Note that magickly will pass the value of the query param to the block as a single string.

See the [Dragonfly documentation](http://markevans.github.com/dragonfly/file.GeneralUsage.html#Shortcuts) for more info on "shortcuts", and the [shortcuts.rb](https://github.com/afeld/magickly/blob/master/lib/shortcuts.rb) file for examples.

## Disclaimer

The hosted version of the app ([magickly.jux.com](http://magickly.jux.com)) is a single app instance intended for demonstration purposes - if you are going to be making a large number of API calls to it or would like to use it in production, please let us know :-)

## Contributing to magickly
 
* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it
* Fork the project
* Start a feature/bugfix branch
* Commit and push until you are happy with your contribution
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
* Please try not to mess with the Rakefile, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

## Credits

Created by [Aidan Feldman](http://www.aidanfeldman.com) at [Jux.com](http://www.jux.com).  Thanks to [Mark Evans](https://github.com/markevans) for all his hard work on [Dragonfly](http://markevans.github.com/dragonfly/).



HISTORY.md
This gem respects [semantic versioning](http://semver.org/).

# 1.4.0 (1/26/12)

[commits](https://github.com/afeld/magickly/compare/v1.3.0...v1.4.0)

* SECURITY: don't pass cookies to image request (tolsen)
* removed async support (for now) - needs more extensive testing
* added /q syntax as alternative to query params (tolsen)
* added JCN effect

# 1.2.1 (12/21/11)

[commits](https://github.com/afeld/magickly/compare/v1.2.0...v1.2.1)

* made compatible w/ Dragonfly v0.9.8 (tolsen)

# 1.3.0 (9/5/11) - YANKED

[commits](https://github.com/afeld/magickly/compare/v1.2.0...v1.3.0)

* added asynchronous support when using Ruby 1.9 or JRuby
* reduced gem dependencies

# 1.2.0 (6/28/11)

[commits](https://github.com/afeld/magickly/compare/v1.1.3...v1.2.0)

* add /analyze and /analyze/:analyzer endpoints
* fix for negative values for brightness_contrast
* validation & specs for existing shortcuts
* added shortcuts:
    * color_palette (analyzer)
    * cross_process
    * glow
    * halftone
    * lomo
    * saturation
    * two_color

# 1.1.3 (5/26/11)

[commits](https://github.com/afeld/magickly/compare/v1.1.2...v1.1.3)

* make tilt-shift faster
* add brightness_contrast shortcut

# 1.1.2 (5/16/11)

[commits](https://github.com/afeld/magickly/compare/v1.1.1...v1.1.2)

* remove rack-cache - this is not the responsibility of the gem
* use NewRelic, if configured
* ignore params that do not correspond to jobs or processor methods
* get job/processor commands from Dragonfly (new in v0.9.1)
* enable tilt_shift shortcut (possible after `convert` argument order change in Dragonfly v0.9.0)

# 1.1.1 (4/28/11)

[commits](https://github.com/afeld/magickly/compare/v1.1.0...v1.1.1)

* cross-compatibility fix for Ruby 1.8/1.9 regarding the checking of available methods from the Processor
* Addressable now required properly when included as a gem

# 1.1.0 (4/26/11)

[commits](https://github.com/afeld/magickly/compare/v1.0.0...v1.1.0)

* make underlying Dragonfly application accessible for external configuration
* added some example Dragonfly shortcuts
* major refactoring: namespacing inside of Magickly module (should be entirely backwards compatible)

# 1.0.0 (4/19/11)

* initial release


lib/magickly/convert/factories.rb
module Magickly

  extend Magickly::ConvertHelpers

  LOMO_MASK_PATH = File.join(File.dirname(__FILE__), '..', '..', 'images', 'lomo_mask.png')

  THUMB_FOCUS_GEOMETRY = /^(\d+)x(\d+)#(\d+(\.\d+)?),(\d+(\.\d+)?)$/

  add_simple_convert_factory :auto_orient, "-auto-orient"

  add_simple_convert_factory(:brightness_contrast,
                             "-brightness-contrast",
                             true,
                             /^-?\d+%?(x-?\d+%?)?$/,
                             "argument must be of the format '<int>[%]x<int>[%]'")

  add_convert_factory :color_palette_swatch do |c|
    c.convert_args do |count, convert|
      count = Magickly::DEFAULT_PALETTE_COLOR_COUNT if count == 'true'
      "-resize 600x600 -colors #{count} -unique-colors -scale 10000%"
    end

    c.identity do |count, convert|
      convert.pre_identify.merge :width => 600, :height => 600, :format => :gif
    end

    c.force_format = :gif
  end

  # thanks to http://www.photoshopsupport.com/tutorials/or/cross-processing.html
  add_simple_convert_factory :cross_process, "-channel Red -sigmoidal-contrast 6,50% -channel Blue -level 25%\\! -channel Green -sigmoidal-contrast 5,45% \\( +clone +matte -fill yellow -colorize 4% \\) -compose overlay -composite"

  add_simple_convert_factory :flip, "-flip"

  add_simple_convert_factory :flop, "-flop"

  # thanks to Fred Weinhaus (http://www.fmwconcepts.com/imagemagick)
  add_convert_factory :glow do |c|
    c.convert_args do |value, convert|
      if value == 'true'
        amount = 1.2
        softening = 20
      elsif value =~ /^(\d+\.\d+?),(\d+)$/ && $1.to_f >= 1.0 && $2.to_i >= 0
        amount = $1
        softening = $2
      else
        raise ArgumentError, "args must be of the form <amount(float)>,<softening(int)>"
      end
      
      "\\( +clone -evaluate multiply #{amount} -blur 0x#{softening} \\) -compose plus -composite"
    end
  end

  add_simple_convert_factory :grayscale, "-colorspace Gray"

  add_simple_convert_factory :greyscale, "-colorspace Gray"

  # thanks to http://www.melissaevans.com/tutorials/pop-art-inspired-by-lichtenstein
  add_convert_factory :halftone do |c|
    c.convert_args do |threshold, convert|
      threshold = 50 if threshold == 'true'
      "-white-threshold #{threshold.to_i}% -gaussian-blur 2 -ordered-dither 8x1"
    end
  end

  # thanks to Jesse Chan-Norris - http://jcn.me/
  add_simple_convert_factory :jcn, "-colorspace Gray -white-threshold 99% -gaussian-blur 2 -ordered-dither 8x1"

  ## thanks to https://github.com/soveran/lomo
  add_convert_factory :lomo do |c|
    c.convert_args do |modulate_params, convert|
      if modulate_params == 'true'
        modulate_params = '100,150'
      elsif modulate_params !~ /^\d+,\d+$/
        raise ArgumentError, "modulate_params must be of the format '<int>,<int>'"
      end

      width, height = convert.pre_identify.values_at :width, :height
      
      "\\( +clone -unsharp 1 -contrast -contrast -modulate #{modulate_params} \\( #{LOMO_MASK_PATH} -resize #{width}x#{height}\\! \\) -compose overlay -composite \\) -compose multiply -composite"
    end
  end

  add_convert_factory :resize do |c|
    c.convert_args do |geometry, convert|
      "-resize #{geometry}"
    end

    c.identity do |geometry, convert|
      identity = convert.pre_identify
      identity.merge resized_dimensions(identity[:width], identity[:height], geometry)
    end
  end

  add_simple_convert_factory :resize_with_blur, "-filter Gaussian -resize", true

  add_simple_convert_factory :rotate, "-rotate", true

  add_convert_factory :saturation do |c|
    c.convert_args do |percentage, convert|
      raise ArgumentError, "percentage must be a positive integer" unless percentage =~ /^\d+$/
      "-modulate 100,#{percentage}"
    end
  end

  # Adapted from dragonfly
  add_convert_factory :thumb do |c|
    c.convert_args do |geometry, convert|
      case geometry
      when Dragonfly::ImageMagick::Processor::RESIZE_GEOMETRY
        "-resize #{geometry}"
      when Dragonfly::ImageMagick::Processor::CROPPED_RESIZE_GEOMETRY
        resize_and_crop_args :width => $1, :height => $2, :gravity => $3
      when Dragonfly::ImageMagick::Processor::CROP_GEOMETRY
        crop_args(
                  :width => $1,
                  :height => $2,
                  :x => $3,
                  :y => $4,
                  :gravity => $5
                  )
      when THUMB_FOCUS_GEOMETRY
        identity = convert.pre_identify
        
        resize_and_crop_with_focus_args(
          container_width: $1.to_i,
          container_height: $2.to_i,
          focus_x: $3.to_f / 100,
          focus_y: $5.to_f / 100,
          orig_width: identity[:width],
          orig_height: identity[:height],
        )
      else raise ArgumentError, "Didn't recognise the geometry string #{geometry}"
      end
    end

    c.identity do |geometry, convert|
      identity = convert.pre_identify
      width, height = identity.values_at :width, :height
      case geometry
      when Dragonfly::ImageMagick::Processor::RESIZE_GEOMETRY
        identity.merge resized_dimensions(width, height, geometry)
      when Dragonfly::ImageMagick::Processor::CROPPED_RESIZE_GEOMETRY, Dragonfly::ImageMagick::Processor::CROP_GEOMETRY, THUMB_FOCUS_GEOMETRY
        identity.merge :width => $1, :height => $2
      else raise ArgumentError, "Didn't recognise the geometry string #{geometry}"
      end
    end
  end

  add_convert_factory :tilt_shift do |c|
    c.convert_args do |coefficients, convert|
      coefficients = '4,-4,1' if coefficients == 'true'
      raise ArgumentError, "coefficients must be of the format '<decimal>,<decimal>,<decimal>'" unless coefficients =~ /^(-?\d+(\.\d+)?,){2}-?\d+(\.\d+)?$/
    
      # note: can be made faster by decreasing sigma passed to option:compose:args
      "\\( +clone -sparse-color Barycentric '0,0 black 0,%h white' -function polynomial #{coefficients} \\) -compose Blur -set option:compose:args 8 -composite"
    end
  end

  # thanks to Fred Weinhaus (http://www.fmwconcepts.com/imagemagick)
  add_simple_convert_factory :two_color, "-background black -flatten +matte +dither -colors 2 -colorspace gray -contrast-stretch 0"
  
end
